### PR TITLE
Add private attributes support

### LINKED ISSUE #655: Question: add private attribute
I have a use case that I'd to add an attribute when initialising the instance which is not part of the model, thus should not be validated. Is that possible?

Here's a practical example:

```py
from pydantic import BaseModel
from datetime import datetime

class Test(BaseModel):
    a: int

class TestExtra(BaseModel):
    a: int

    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self._processed_at = datetime.utcnow()

test = {"a": 1}
Test(**test)
TestExtra(**test)  # ValueError: "TestExtra" object has no field "_processed_at"
```

### PR DESCRIPTION
**Pull request by MrMrRobat**
## Change Summary
Private attributes declared as regular fields, but always start with underscore and `PrivateAttr` is used instead of `Field`. Upon class creation they added in `__slots__` and `Model.__private_attributes__` for faster lookups and support for default values
Default values declared just like fields: `_attr[: annotation] = default_value` or `_attr = PrivateAttr(default_value)`

## Example from docs
```py
from datetime import datetime
from random import randint

from pydantic import BaseModel, PrivateAttr


class TimeAwareModel(BaseModel):
    _processed_at: datetime = PrivateAttr(default_factory=datetime.now)
    _secret_value: str = PrivateAttr()

    def __init__(self, **data):
        super().__init__(**data)
        self._secret_value = randint(1, 5)


m = TimeAwareModel()
print(m._processed_at)
print(m._secret_value)

```

## Other changes
Replaced `copy.deepcopy()` usage with new `pydantic.utils.smart_deepcopy()` for faster immutable values handling

## Related issue number
Closes #655

## Checklist
* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `changes/<pull request or issue id>-<github username>.md` file added describing change
  (see [changes/README.md](https://github.com/samuelcolvin/pydantic/blob/master/changes/README.md) for details)

### CODE REVIEW

**Comment by MrMrRobat** [comment]
@samuelcolvin, seems like tests merged from `master` were broken

**Comment by Benoss** [comment]
I quite like this way of doing it

File: pydantic/main.py [hunk 1]
```diff
@@ -363,11 +373,23 @@ def __setattr__(self, name, value):
...
+        }
 
     def __setstate__(self, state: 'DictAny') -> None:
-        object.__setattr__(self, '__dict__', state['__dict__'])
-        object.__setattr__(self, '__fields_set__', state['__fields_set__'])
+        object_setattr(self, '__dict__', state['__dict__'])
+        object_setattr(self, '__fields_set__', state['__fields_set__'])
+        self._set_private_attributes(state['__private_attributes_values__'], need_copy=False, check=False)
+
+    def _set_private_attributes(self, source: 'DictStrAny', need_copy: bool = True, check: bool = True) -> None:
```
Comment (samuelcolvin) [inline pydantic/main.py:387]: Looks like the two kwargs are always used together, maybe they can be combined into `check_copy` or `from_pickle=False`?

This method might need a docstring too.

File: pydantic/main.py [hunk 2]
```diff
@@ -363,11 +373,23 @@ def __setattr__(self, name, value):
...
+        for name, value in source.items():
+            if value is not Undefined:
+                object_setattr(self, name, smart_deepcopy(value) if need_copy else value)
+            elif check and not hasattr(self, name):
+                raise AttributeError(f'private attribute "{name}" is unset')
```
Comment (samuelcolvin) [inline pydantic/main.py:392]: maybe if check is `True` we should do the check first so the exception message is always the same?

Comment (samuelcolvin) [reply pydantic/main.py:392]: or make it identical here.

File: docs/examples/private_attributes.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+from pydantic import BaseModel, Field
+
+storage = {}
+
+
+def get_new_id():
+    return max(storage) + 1 if storage else 0
+
+
+class BaseStorageModel(BaseModel):
```
Comment (samuelcolvin) [inline docs/examples/private_attributes.py:12]: maybe the `_processed_at` example at the top of #655 is simpler and clearer?

**Review by samuelcolvin — COMMENTED** [review]
sorry it's been so long, I've been really busy.

This looks amazing, but might need to be split into multiple PRs.

File: pydantic/utils.py [hunk 4]
```diff
@@ -477,3 +495,36 @@ def __get__(self, instance: Any, owner: Type[Any]) -> None:
...
+            raise NameError(
+                f'Private attributes "{name}" must not be a valid field name; '
+                f'Use sunder or dunder names, e. g. "_{name}" or "__{name}__"'
+            )
+
+
+Obj = TypeVar('Obj')
+
+
+def smart_deepcopy(obj: Obj) -> Obj:
```
[the remark this answers is not in the document]
Comment (MrMrRobat) [reply pydantic/utils.py:521]: You're right, separate PR is more sutable for this change.

> Or just works in all the cases we know of? 

Good catch. I think implementing faster behaviour only for built-in collections and immutable types is a way to go.

File: pydantic/main.py [hunk 5]
```diff
@@ -363,11 +373,23 @@ def __setattr__(self, name, value):
...
+            '__fields_set__': self.__fields_set__,
+            '__private_attributes_values__': {k: getattr(self, k, Undefined) for k in self.__private_attributes__},
+        }
 
     def __setstate__(self, state: 'DictAny') -> None:
-        object.__setattr__(self, '__dict__', state['__dict__'])
-        object.__setattr__(self, '__fields_set__', state['__fields_set__'])
+        object_setattr(self, '__dict__', state['__dict__'])
+        object_setattr(self, '__fields_set__', state['__fields_set__'])
+        self._set_private_attributes(state['__private_attributes_values__'], need_copy=False, check=False)
```
Comment (MrMrRobat) [inline pydantic/main.py:385]: @samuelcolvin, maybe we should use `dict.get()` here, so unpickling models dumped with older versions of `pydantic` would work?
```suggestion
        self._set_private_attributes(state.get('__private_attributes_values__', {}), need_copy=False, check=False)
```

File: pydantic/main.py [hunk 6]
```diff
@@ -205,15 +202,19 @@ def __new__(mcs, name, bases, namespace, **kwargs):  # noqa C901
...
-
         pre_root_validators, post_root_validators = [], []
-        for base in reversed(bases):
-            if _is_base_model_class_defined and issubclass(base, BaseModel) and base != BaseModel:
-                fields.update(deepcopy(base.__fields__))
+        slots: Union[str, Tuple[str, ...]] = namespace.get('__slots__', ())
+        private_attributes: Dict[str, Any] = dict.fromkeys((slots,) if isinstance(slots, str) else slots, Undefined)
+        validate_private_attributes(private_attributes)
+
+        for base in reversed(bases) if _is_base_model_class_defined else ():
```
[the remark this answers is not in the document]
Comment (MrMrRobat) [reply pydantic/main.py:210]: Oof, unfortunately I don't remember. Should've commented it out 🤦 . But from what I see, now it includes `BaseModel` and it didn't before. I'll try to remember the reason I did it this way and either put explaning comment, or change the line back.

**Review by MrMrRobat — COMMENTED** [review]
@samuelcolvin, thanks for review, I answered some comments and think I'll be able to split PR on the next weekend.

File: pydantic/main.py (hunk 5 above)
Comment (samuelcolvin) [reply pydantic/main.py:385]: Yes makes sense.

File: pydantic/utils.py [hunk 7]
```diff
@@ -40,14 +41,31 @@
...
     'Representation',
     'GetterDict',
     'ValueItems',
     'version_info',  # required here to match behaviour in v1.3
     'ClassAttribute',
+    'ROOT_KEY',
 )
 
+ROOT_KEY = '__root__'
+IMMUTABLE_NON_COLLECTIONS_TYPES: AbstractSet[Type[Any]] = {
```
[the remark this answers is not in the document]
Comment (samuelcolvin) [reply pydantic/utils.py:57]: ok

**Review by samuelcolvin — COMMENTED** [review]
Most things look good here, met me know if I've missed any questions.

**Comment by samuelcolvin** [comment]
automatically generating `__slots__` would be great, I think this would make it much easier to use.

**Comment by MrMrRobat** [comment]
What do you think about `PrivateField`? Should it subclass regular field class, be an independent class, or it's better to use `Field(private=True)`?

**Comment by samuelcolvin** [comment]
I think probably easiest if it's a regular class.

I think we should keep it separate from `Field` because the point is that these attributes are NOT fields.

File: pydantic/main.py (hunk 2 above)
Comment (MrMrRobat) [reply pydantic/main.py:392]: Didn't get this one, could you explain?

File: pydantic/main.py [hunk 8]
```diff
@@ -339,17 +357,18 @@ def __init__(__pydantic_self__, **data: Any) -> None:
...
             raise validation_error
-        object.__setattr__(__pydantic_self__, '__dict__', values)
-        object.__setattr__(__pydantic_self__, '__fields_set__', fields_set)
+        object_setattr(__pydantic_self__, '__dict__', values)
+        object_setattr(__pydantic_self__, '__fields_set__', fields_set)
+        __pydantic_self__._set_default_private_attributes(__pydantic_self__.__private_attributes__)
 
     @no_type_check
-    def __setattr__(self, name, value):
+    def __setattr__(self, name, value):  # noqa: C901 (ignore complexity)
```
Comment (MrMrRobat) [inline pydantic/main.py:368]: had to set `# noqa: C901 (ignore complexity)` here because of today's changes in #1972. Is this the way to go, or some recent changes have to go outside of this method?

Comment (samuelcolvin) [reply pydantic/main.py:368]: i think this is fine.

**Comment by samuelcolvin** [comment]
@MrMrRobat this is looking good, do you think it will be ready to merge today?

I want to get v1.7 released, but I'm happy wait until tomorrow if we can include this?

**Comment by MrMrRobat** [comment]
> do you think it will be ready to merge today?

Yeah, if don't have any comments or suggestions, it's ready :)

File: pydantic/main.py [hunk 9]
```diff
@@ -339,17 +357,18 @@ def __init__(__pydantic_self__, **data: Any) -> None:
...
-            __pydantic_self__.__dict__: Dict[str, Any] = {}
-            __pydantic_self__.__fields_set__: 'SetStr' = set()
         values, fields_set, validation_error = validate_model(__pydantic_self__.__class__, data)
         if validation_error:
             raise validation_error
-        object.__setattr__(__pydantic_self__, '__dict__', values)
-        object.__setattr__(__pydantic_self__, '__fields_set__', fields_set)
+        object_setattr(__pydantic_self__, '__dict__', values)
+        object_setattr(__pydantic_self__, '__fields_set__', fields_set)
+        __pydantic_self__._set_default_private_attributes(__pydantic_self__.__private_attributes__)
```
Comment (samuelcolvin) [inline pydantic/main.py:365]: why is this required?

File: docs/examples/private_attributes_underscore_attrs_are_private.py [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+
+from pydantic import BaseModel
+
+
+class Model(BaseModel):
+    __class_var__: ClassVar[str] = 'class var value'
+    __private_attr__: str = 'private attr value'
+
+    class Config:
+        underscore_attrs_are_private = True
```
Comment (samuelcolvin) [inline docs/examples/private_attributes_underscore_attrs_are_private.py:11]: I this needs documenting in `model_config.md`. Also we need to be clear exactly what form names need to be in.

It looks from this like if you use `PrivateAttr` you can use `_whatever`, but with `underscore_attrs_are_private` you have to use `__whatever__`, but reading the code, I think that's not the case.

File: docs/examples/private_attributes.py [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+from pydantic import BaseModel, PrivateAttr
+
+
+class TimeAwareModel(BaseModel):
+    _processed_at: datetime = PrivateAttr(default_factory=datetime.now)
+    _secret_value: str = PrivateAttr()
+
+    def __init__(self, **data):
+        super().__init__(**data)
+        self._secret_value = randint(1, 5)
```
Comment (samuelcolvin) [inline docs/examples/private_attributes.py:13]: maybe add a commit that this could also be done with `default_factory`

File: docs/examples/private_attributes_underscore_attrs_are_private.py (hunk 10 above)
Comment (samuelcolvin) [reply docs/examples/private_attributes_underscore_attrs_are_private.py:11]: actually you've written it clear in the copy, but still better to use the same format in both code examples.

Is there anything special about `__whatever__` or is it just that it starts with an underscore, like `_whatever`?

Comment (samuelcolvin) [reply docs/examples/private_attributes_underscore_attrs_are_private.py:11]: Since I think according to some PEP we're not supposed to use dunder methods, unless they're official to python, maybe leave them out of the example and just mention them in the text?

File: pydantic/main.py [hunk 12]
```diff
@@ -123,6 +126,7 @@ class BaseConfig:
     json_loads: Callable[[str], Any] = json.loads
     json_dumps: Callable[..., str] = json.dumps
     json_encoders: Dict[Type[Any], AnyCallable] = {}
+    underscore_attrs_are_private = False
```
Comment (samuelcolvin) [inline pydantic/main.py:129]:
```suggestion
    underscore_attrs_are_private: bool = False
```
I know it's obvious, but best to be explicit.

**Review by samuelcolvin — COMMENTED** [review]
looking good, just a few things to clarify.

File: docs/examples/private_attributes.py (hunk 11 above)
Comment (MrMrRobat) [reply docs/examples/private_attributes.py:13]: you mean comment?

File: pydantic/main.py (hunk 9 above)
Comment (MrMrRobat) [reply pydantic/main.py:365]: what do you mean exactly?

File: docs/examples/private_attributes.py (hunk 11 above)
Comment (samuelcolvin) [reply docs/examples/private_attributes.py:14]: _yes_

File: pydantic/main.py (hunk 9 above)
Comment (samuelcolvin) [reply pydantic/main.py:365]: Can we avoid this line since this is the critical path in terms of performance?

Comment (MrMrRobat) [reply pydantic/main.py:365]: It sets defaults of `__private_attributes__`, without this line `default` not `default_factoru` wouldn't work.
What's critical here in terms of performance? Function call itself or executing code of the function?

Comment (samuelcolvin) [reply pydantic/main.py:365]: Ok, I guess we need it.

Initialising/validating models is the critical path, I don't want to slow that down unless we absolutely have to.

File: pydantic/main.py [hunk 13]
```diff
@@ -385,11 +404,24 @@ def __setattr__(self, name, value):
...
     def __setstate__(self, state: 'DictAny') -> None:
-        object.__setattr__(self, '__dict__', state['__dict__'])
-        object.__setattr__(self, '__fields_set__', state['__fields_set__'])
+        object_setattr(self, '__dict__', state['__dict__'])
+        object_setattr(self, '__fields_set__', state['__fields_set__'])
+        for name, value in state.get('__private_attribute_values__', {}).items():
+            if value is not Undefined:
+                object_setattr(self, name, value)
+
+    def _set_default_private_attributes(self, source: Dict[str, PrivateAttr]) -> None:
```
Comment (samuelcolvin) [inline pydantic/main.py:420]:
```suggestion
    def _init_private_attributes(self, source: Dict[str, PrivateAttr]) -> None:
```
maybe?

More concise and explains it well I think.

**Review by samuelcolvin — COMMENTED** [review]
I think this is looking amazing, thank you @MrMrRobat.

I've tried it myself and i think it'll be extremely useful.

I have one small proposed change in name, but only other question is on `model.copy()` which currently doesn't copy private attributes, but instead initialises from the default or default factory. I think I would expect `copy()` to (deep?)copy the private attributes, a bit like `__setstate__`.

What do you think?

**Comment by samuelcolvin** [comment]
you need to reference the new name, sorry I should have made that clear in my change proposal.

Will you change it or will I?

**Comment by MrMrRobat** [comment]
>  I think I would expect copy() to (deep?)copy the private attributes, a bit like __setstate__.

Yeah, you right. Working on it.

**Comment by MrMrRobat** [comment]
> Working on it.

Done

**Comment by samuelcolvin** [comment]
awesome, thank you.

I'm really looking forward to using this.

**Comment by samuelcolvin** [comment]
Question on this in https://github.com/pydantic/pydantic/discussions/4518.
